yktrace/src/hwt/mapper.rs

```diff
@@ -127,8 +128,25 @@ impl HWTMapper {
                     }
                 }
             } else {
-                ret_irblocks.extend(irblocks);
+                for irblock in irblocks.into_iter() {
+                    // A basic block may map to >1 *machine* basic block. If we see the same block
+                    // repeated, then we need to check if we re-executing the same basic block or
+                    // if we are executing next machine basic block for that same block. In the
+                    // latter case, we mustn't record the block in the trace again.
+                    let mut append = true;
```

Comment:
The logic in lines 136-147 is a bit hard to follow. I *think* it's equivalent to something like :

```
if ret_irblocks.last() == Some(irblock) && prev_block.map(|x| x.first_instr()) != Some(block.first_instr()) {
    ret_irblocks.push(irblock);
}
```

Reply:
```
if ret_irblocks.last() == Some(irblock)
```

`if let`?

You can't combine `if let` with `&&` hence this logic being a bit messy.

I'll have a go at simplifying it a bit later.

Reply:
I deliberately didn't use `if let` because `if let` doesn't compose syntactically.